rules/opentitan.bzl

```diff
@@ -0,0 +1,203 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+# TODO(drewmacrae) this should be in rules_cc
+load("//rules:bugfix.bzl", "find_cc_toolchain")
+
+"""Rules to build OpenTitan for the RiscV target"""
+
+OPENTITAN_CPU = "@bazel_embedded//constraints/cpu:riscv32"
+OPENTITAN_PLATFORM = "@bazel_embedded//platforms:opentitan_rv32imc"
+
+_targets_compatible_with = {
+    OPENTITAN_PLATFORM: [OPENTITAN_CPU],
+}
+
+def _platforms_transition_impl(settings, attr):
+    return {"//command_line_option:platforms": attr.platform}
+
+_platforms_transition = transition(
+    implementation = _platforms_transition_impl,
+    inputs = [],
+    outputs = ["//command_line_option:platforms"],
+)
+
+def _elf_to_binary(ctx):
+    cc_toolchain = find_cc_toolchain(ctx)
+    outputs = []
+    for src in ctx.files.srcs:
+        binary = ctx.actions.declare_file("{}.{}".format(src.basename, ctx.attr.suffix))
+        outputs.append(binary)
+        ctx.actions.run(
+            outputs = [binary],
+            inputs = [src] + cc_toolchain.all_files.to_list(),
+            arguments = [
+                "--output-target",
+                ctx.attr.format,
+                src.path,
+                binary.path,
+            ],
+            executable = cc_toolchain.objcopy_executable,
+        )
+    return [DefaultInfo(files = depset(outputs), data_runfiles = ctx.runfiles(files = outputs))]
+
+elf_to_binary = rule(
+    implementation = _elf_to_binary,
+    cfg = _platforms_transition,
+    attrs = {
+        "srcs": attr.label_list(allow_files = True),
```

Comment:
Is a list of labels actually really correct? I don't think relinking multiple targets' binaries in one rule makes sense.

(I would also call this `bin` or similar; the input is a build artifact, not a source file.)

Reply:
"Relinking" isn't really what's happening.  Its more of a transformation from one binary object format to another (ie: ELF to COFF, MachO, or flat binary, etc).

It's common convention for bazel rules to name their inputs `srcs` (For example, `filegroup` and `genrule` name their inputs as `srcs`.  Its the "source" to the rule, not "source code").

As to whether a list of labels or a single label is correct, that's a bit harder to judge.  The implementation of the rule just wraps a `for` loop around this, so it turns into an invocation of `objcopy` per item in the `srcs` list... which will almost always only be one label.

Reply:
I don't have a strong feeling; I have seen rules that go for a single label in this kind of case (I would have gone with single label).